File: app/src/main/java/de/txserver/slickupnp/upnp/CustomContentBrowseActionCallback.java
package de.txserver.slickupnp.upnp;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import org.fourthline.cling.android.AndroidUpnpService;
import org.fourthline.cling.model.action.ActionException;
import org.fourthline.cling.model.action.ActionInvocation;
import org.fourthline.cling.model.message.UpnpResponse;
import org.fourthline.cling.model.meta.Service;
import org.fourthline.cling.model.types.ErrorCode;
import org.fourthline.cling.support.contentdirectory.callback.Browse;
import org.fourthline.cling.support.model.BrowseFlag;
import org.fourthline.cling.support.model.DIDLContent;
import org.fourthline.cling.support.model.DIDLObject;
import org.fourthline.cling.support.model.SortCriterion;
import org.fourthline.cling.support.model.container.Container;
import org.fourthline.cling.support.model.item.Item;

import java.net.URI;
import java.util.ArrayList;

import de.txserver.slickupnp.R;
import de.txserver.slickupnp.app.SlickUPnP;
import de.txserver.slickupnp.helper.ItemModel;
import de.txserver.slickupnp.helper.MimeTypeMap;

public class CustomContentBrowseActionCallback extends Browse {

    private Context context;
    private SharedPreferences prefs;

    private ContentDirectoryBrowseCallbacks callbacks;
    private AndroidUpnpService androidUpnpService;
    private Service service;
    private String id;
    private long firstResult;

    public CustomContentBrowseActionCallback(Context context, ContentDirectoryBrowseCallbacks callbacks, AndroidUpnpService androidUpnpService, Service service, String id) {
        this(context, callbacks, androidUpnpService, service, id, 0L);
    }

    public CustomContentBrowseActionCallback(Context context, ContentDirectoryBrowseCallbacks callbacks, AndroidUpnpService androidUpnpService, Service service, String id, long firstResult) {
        super(service, id, BrowseFlag.DIRECT_CHILDREN, "*", firstResult, 99999L,
                new SortCriterion(true, "dc:title"));

        this.callbacks = callbacks;
        this.androidUpnpService = androidUpnpService;
        this.service = service;
        this.id = id;
        this.firstResult = firstResult;

        this.context = context;
        prefs = SlickUPnP.getInstance().getSharedPref();
    }

    private ItemModel createItemModel(DIDLObject item) {

        ItemModel itemModel = new ItemModel(context.getResources(),
                R.drawable.ic_folder_black_24px, service, item);

        URI usableIcon = item.getFirstPropertyValue(DIDLObject.Property.UPNP.ICON.class);
        if (usableIcon == null || usableIcon.toString().isEmpty()) {
            usableIcon = item.getFirstPropertyValue(DIDLObject.Property.UPNP.ALBUM_ART_URI.class);
        }
        if (usableIcon != null)
            itemModel.setIconUrl(usableIcon.toString());

        if (item instanceof Item) {
            String mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(MimeTypeMap.getFileExtensionFromUrl(itemModel.getUrl()));

            if (mimeType.contains("video")) {
                itemModel.setIcon(R.drawable.ic_file_video_black_24px);

            } else if (mimeType.contains("audio")) {
                itemModel.setIcon(R.drawable.ic_file_music_black_24px);

            } else if (mimeType.contains("image")) {
                itemModel.setIcon(R.drawable.ic_file_image_black_24px);

            } else {
                itemModel.setIcon(R.drawable.ic_file_black_24px);
            }

            if (prefs.getBoolean("settings_hide_file_icons", false))
                itemModel.setHideIcon(true);

            if (prefs.getBoolean("settings_show_extensions", false))
                itemModel.setShowExtension(true);
        }

        return itemModel;
    }

    @Override
    public void received(final ActionInvocation actionInvocation, final DIDLContent didl) {

        ArrayList<ItemModel> items = new ArrayList<ItemModel>();
        Long totalMatches;
        Long numberReturned;

        try {
            for (Container childContainer : didl.getContainers())
                items.add(createItemModel(childContainer));

            for (Item childItem : didl.getItems())
                items.add(createItemModel(childItem));

            totalMatches = Long.parseLong(actionInvocation.getOutput("TotalMatches").toString());
//                numberReturned = Long.parseLong(actionInvocation.getOutput("NumberReturned").toString());

            if (firstResult < 9999L && firstResult + items.size() < totalMatches) {

                androidUpnpService.getControlPoint().execute(
                        new CustomContentBrowseActionCallback(context, callbacks, androidUpnpService, service,
                                id, firstResult + items.size()));


            }

            if (firstResult == 0) {
                callbacks.onDisplayItems(items);
            } else {
                callbacks.onDisplayAddItems(items);
            }

        } catch (Exception ex) {
            actionInvocation.setFailure(new ActionException(
                    ErrorCode.ACTION_FAILED,
                    "Can't create list childs: " + ex, ex));
            failure(actionInvocation, null, ex.getMessage());
        }
    }

    @Override
    public void updateStatus(Status status) {

    }

    @Override
    public void failure(ActionInvocation invocation, UpnpResponse response, String s) {
        callbacks.onDisplayItemsError(createDefaultFailureMessage(invocation, response));
    }
}

File: app/src/main/java/de/txserver/slickupnp/upnp/CustomContentBrowseActionCallback.kt
package de.txserver.slickupnp.upnp

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager

import org.fourthline.cling.android.AndroidUpnpService
import org.fourthline.cling.model.action.ActionException
import org.fourthline.cling.model.action.ActionInvocation
import org.fourthline.cling.model.message.UpnpResponse
import org.fourthline.cling.model.meta.Service
import org.fourthline.cling.model.types.ErrorCode
import org.fourthline.cling.support.contentdirectory.callback.Browse
import org.fourthline.cling.support.model.BrowseFlag
import org.fourthline.cling.support.model.DIDLContent
import org.fourthline.cling.support.model.DIDLObject
import org.fourthline.cling.support.model.SortCriterion
import org.fourthline.cling.support.model.container.Container
import org.fourthline.cling.support.model.item.Item

import java.net.URI
import java.util.ArrayList

import de.txserver.slickupnp.R
import de.txserver.slickupnp.app.SlickUPnP
import de.txserver.slickupnp.helper.ItemModel
import de.txserver.slickupnp.helper.MimeTypeMap

class CustomContentBrowseActionCallback @JvmOverloads constructor(private val context: Context, private val callbacks: ContentDirectoryBrowseCallbacks, private val androidUpnpService: AndroidUpnpService, private val service: Service<*, *>, private val id: String, private val firstResult: Long = 0L) : Browse(service, id, BrowseFlag.DIRECT_CHILDREN, "*", firstResult, 99999L, SortCriterion(true, "dc:title")) {
    private val prefs: SharedPreferences

    init {
        prefs = SlickUPnP.instance!!.getSharedPref()
    }

    private fun createItemModel(item: DIDLObject): ItemModel {

        val itemModel = ItemModel(context.resources,
                R.drawable.ic_folder_black_24px, service, item)

        var usableIcon: URI? = item.getFirstPropertyValue(DIDLObject.Property.UPNP.ICON::class.java)
        if (usableIcon == null || usableIcon.toString().isEmpty()) {
            usableIcon = item.getFirstPropertyValue(DIDLObject.Property.UPNP.ALBUM_ART_URI::class.java)
        }
        if (usableIcon != null)
            itemModel.iconUrl = usableIcon.toString()

        if (item is Item) {
            val mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(MimeTypeMap.getFileExtensionFromUrl(itemModel.url))

            if (mimeType!!.contains("video")) {
                itemModel.icon = R.drawable.ic_file_video_black_24px

            } else if (mimeType.contains("audio")) {
                itemModel.icon = R.drawable.ic_file_music_black_24px

            } else if (mimeType.contains("image")) {
                itemModel.icon = R.drawable.ic_file_image_black_24px

            } else {
                itemModel.icon = R.drawable.ic_file_black_24px
            }

            if (prefs.getBoolean("settings_hide_file_icons", false))
                itemModel.hideIcon = true

            if (prefs.getBoolean("settings_show_extensions", false))
                itemModel.setShowExtension(true)
        }

        return itemModel
    }

    override fun received(actionInvocation: ActionInvocation<*>, didl: DIDLContent) {

        val items = ArrayList<ItemModel>()
        val totalMatches: Long?

        try {
            for (childContainer in didl.containers)
                items.add(createItemModel(childContainer))

            for (childItem in didl.items)
                items.add(createItemModel(childItem))

            totalMatches = java.lang.Long.parseLong(actionInvocation.getOutput("TotalMatches").toString())

            if (firstResult < 9999L && firstResult + items.size < totalMatches) {

                androidUpnpService.controlPoint.execute(
                        CustomContentBrowseActionCallback(context, callbacks, androidUpnpService, service,
                                id, firstResult + items.size))


            } else {
                callbacks.setShowRefreshing(false)
            }

            if (firstResult == 0L) {
                callbacks.onDisplayItems(items)
            } else {
                callbacks.onDisplayAddItems(items)
            }

        } catch (ex: Exception) {
            actionInvocation.failure = ActionException(
                    ErrorCode.ACTION_FAILED,
                    "Can't create list childs: $ex", ex)
            failure(actionInvocation, null, ex.message!!)
        }

    }

    override fun updateStatus(status: Browse.Status) {

    }

    override fun failure(invocation: ActionInvocation<*>, response: UpnpResponse?, s: String) {
        callbacks.onDisplayItemsError(createDefaultFailureMessage(invocation, response))
        callbacks.setShowRefreshing(false)
    }
}
